en/lessons/basics/functions.md

```diff
@@ -1,425 +1,83 @@
 
-Let's look at a basic example:
-
-```elixir
-iex> sum = fn (a, b) -> a + b end
-iex> sum.(2, 3)
-5
 ```
+   Structure of a lambda
 
-### The & Shorthand
-
-Using anonymous functions is such a common practice in Elixir there is shorthand for doing so:
-
-```elixir
-iex> sum = &(&1 + &2)
-iex> sum.(2, 3)
-5
-```
-
-As you probably already guessed, in the shorthand version our parameters are available to us as `&1`, `&2`, `&3`, and so on.
-
-## Pattern Matching
-
-Pattern matching isn't limited to just variables in Elixir, it can be applied to function signatures as we will see in this section.
-
-Elixir uses pattern matching to check through all possible match options and select the first matching option to run:
-
-```elixir
-iex> handle_result = fn
-...>   {:ok, result} -> IO.puts "Handling result..."
-...>   {:ok, _} -> IO.puts "This would be never run as previous will be matched beforehand."
-...>   {:error} -> IO.puts "An error has occurred!"
-...> end
-
-iex> some_result = 1
-1
-iex> handle_result.({:ok, some_result})
-Handling result...
-:ok
-iex> handle_result.({:error})
-An error has occurred!
+   \x -> x
+    │ │  └─ Body of the lambda
+    │ └─ Separator between head and body of the lambda
+    └─ Declaration of the input bindigs
 ```
 
-## Named Functions
-
-We can define functions with names so we can easily refer to them later.
-Named functions are defined within a module using the `def` keyword .
-We'll learn more about Modules in the next lessons, for now we'll focus on the named functions alone.
+And incidentally, we have also seen one of the most primordial functions : The identity function, or `id`.  
+This function is very simple, as it returns precisely the argument that we pass to it.
 
-Functions defined within a module are available to other modules for use.
-This is a particularly useful building block in Elixir:
-
-```elixir
-defmodule Greeter do
-  def hello(name) do
-    "Hello, " <> name
-  end
-end
-
-iex> Greeter.hello("Sean")
-"Hello, Sean"
-```
+## Named functions
 
-If our function body only spans one line, we can shorten it further with `do:`:
+However, one can only go so far with anonymous functions. This is why we can give names to our functions:
 
-```elixir
-defmodule Greeter do
-  def hello(name), do: "Hello, " <> name
-end
-```
-
-Armed with our knowledge of pattern matching, let's explore recursion using named functions:
-
-```elixir
-defmodule Length do
-  def of([]), do: 0
-  def of([_ | tail]), do: 1 + of(tail)
-end
-
-iex> Length.of []
-0
-iex> Length.of [1, 2, 3]
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> addOne x = x + 1
+ghci> addOne 2
 3
 ```
 
-### Function Naming and Arity
+## Higher-order functions
 
-We mentioned earlier that functions are named by the combination of given name and arity (number of arguments).
-This means you can do things like this:
+We can now see a third way of using functions, which is the composition of both anonymous and named functions:
```

Comment:
Although higher-order functions can compose anonymous and named functions, I'm not sure it's the best way of defining them.

Reply:
I think what I wanted is said more clearly here: https://github.com/haskellfoundation/HaskellSchool/pull/24#issuecomment-873465708.